actionpack/lib/action_dispatch/journey/formatter.rb

```diff
@@ -17,43 +17,27 @@ def initialize(routes)
 
       def generate(name, options, path_parameters, parameterize = nil)
         constraints = path_parameters.merge(options)
-        missing_keys = nil
+        matched_route = nil
 
         match_route(name, constraints) do |route|
-          parameterized_parts = extract_parameterized_parts(route, options, path_parameters, parameterize)
+          matched_route = MatchedRoute.new(route, constraints, options, path_parameters, parameterize)
 
-          # Skip this route unless a name has been provided or it is a
-          # standard Rails route since we can't determine whether an options
-          # hash passed to url_for matches a Rack application or a redirect.
-          next unless name || route.dispatcher?
+          next if matched_route.invalid?
 
-          missing_keys = missing_keys(route, parameterized_parts)
-          next if missing_keys && !missing_keys.empty?
-          params = options.dup.delete_if do |key, _|
-            parameterized_parts.key?(key) || route.defaults.key?(key)
-          end
-
-          defaults       = route.defaults
-          required_parts = route.required_parts
-
-          route.parts.reverse_each do |key|
-            break if defaults[key].nil? && parameterized_parts[key].present?
-            next if parameterized_parts[key].to_s != defaults[key].to_s
-            break if required_parts.include?(key)
+          return matched_route.generate
+        end
 
-            parameterized_parts.delete(key)
+        message = "No route matches #{Hash[constraints.sort_by { |k, v| k.to_s }].inspect}".dup
```

Comment:
As for `dup` I think it has to do with `frozen_string_literal true`, but I am not sure.
To be honest, I have no idea why `<<` works with `frozen_string_literal true`, I think it should be replaced with `+=`

Reply:
Yeah or generate all the strings and `join` them

Reply:
You are perfectly right about the frozen string literals BTW, see https://freelancing-gods.com/2017/07/27/friendly-frozen-string-literals.html . I guess the existing code is fine unless rewritten to use join then. 

Hmm. I wonder if it is necessary to dup it when the string is created with interpolation - it can't be preallocated then in any case.

Reply:
Unfortunately, `sort` raises an error when both `String` and `Symbol` values present, so I will keep `sort_by`, but will use `_` instead of `v` and `to_h` instead of `Hash[...]`